[우아한9조대] 1주차 Facebook 기능 구현

eunsukko opened the pull request:
안녕하세요 리뷰어님 우아한9조대입니다!
팀 프로젝트 1주차 코드 리뷰 요청드립니다.
front 부분은 따로 Pull Request할 예정입니다.

감사합니다.

pobiconan committed:
1주차 Facebook 기능 구현

initial commit

feat : Article Entity 생성

feat : Article save 구현

feat : Article delete 구현

feat : Article update 구현

setting : application.yml 추가

feat : DateTime Entity 생성

setting : rest-assured dependency 추가

feat : Article create api

feat : Article delete api

feat : Article update api

feat : Comment Entity 생성

feat : Comment save 구현

feat : Comment delete, update 구현

feat : Comment create api

feat : Comment delete api

feat : Comment update api

feat : Comment findByArticleId

feat : Article User 매핑

feat : Comment User 매핑

feat : Like Entity

feat : Article 좋아요 기능 구현

feat : Comment 좋아요 기능 구현

feat : Article, Comment 좋아요 여부 확인 기능 추가

refactor : utils 패키지 변경

refactor : 포맷팅

feat : 유저생성 기능 구현

1. User controller
	- implement create [Post]
2. User service
	- implmenet save
3. User entity
	- encrypted password

feat : 유저 조회 기능 구현

1. User controller
   - implement show [Get]
2. User service
   - implmenet findById

feat : 유저 수정 기능 조회

1. User Controller
   - UserApiController, UserWebController
2. User ApiController
   - implement update [Put]
3. User service
   - implmenet update
4. User entity
   - updateModifiableFields

feat : 로그인/로그아웃 기능 구현

1. Login Controller
   - implement login [Post], logout [Get]
2. Login Service
   - implement login
3. SessionUser

feat : 로그인관련 인터셉터 기능 구현

fix : resolve conflict

fix : 코드리뷰 자체피드백 반영

변경사항
- User 관련사항
	- 생성자 private
	- 사용자 체크 로직 추가
	- 비밀번호 체크 로직 추가
- Controller
	- Transactional 클래스에 적용(find 의 경우 readOnly)
	- User 변경시에 session 도 동일하게 수정
- UserOutline 사용 (기존 SessionUser)

fix : 예외에 적절한 기본 메시지 지정

test : ApiControllerTest에 로그인 쿠키 설정

codeanddonuts committed:
feat : 1주차 프론트 엔드 구현

feat: login page layout

fix: female radio button id

fix: resources directory

feat: front-end form validations

feat: further registration form validations

feat: login & registration

refactor: icon color

feat: Article write, Article remove, Comment write

fix : 모든 테스트 통과하도록 수정

- 테스트와 구현 코드의 불일치 부분 수정
	- UserSignupRequest
- profile.html 추가

feat : 유저회원가입 검증 추가

- UserSignUpRequest 변경
	- name
		- 프론트에 맞춰서 lastName, firstName 으로 구분(Assembler 에도 적용)
	- email
		- 이미 존재하는 이메일검증
		- 이메일 양식 검증
	- password
		- password 일치하는지 검증(reconfirmPassword 추가)
		- 비밀번호 정책 추가 (8자 이상의 소문자, 대문자, 숫자, 특수문자의 조합)

- Validator 구현
	- EqualFieldsValidator
		- 두 필드가 동일한지 검증증
	- NotExistsEmailValidator
		- 이미 가입된 이메일인지 검증

fix : 중복된 비밀번호확인 로직 제거

문제: LoginService.login 에서 비밀번호 확인 작업을 두번하고 있었음
- 두가지 방법을 확인해보려고 둘다 작성하고 남겨놓았었음

해결: 한 가지 방식만 남겨놓음으로 해결

refactor: registrations

feat: date formatting

feat: Article editing

phs1116 commented on src/main/java/techcourse/fakebook/controller/ArticleApiController.java:
```diff
@@ -0,0 +1,51 @@
+    @PostMapping
+    public ResponseEntity<ArticleResponse> create(@RequestBody ArticleRequest articleRequest, @SessionUser UserOutline userOutline) {
+        ArticleResponse articleResponse = articleService.save(articleRequest, userOutline);
+        return new ResponseEntity<>(articleResponse, HttpStatus.CREATED);
```
ResponseEntity에는 기본적인 static factory method를 많이 제공합니다.
이를 활용해보면 좋을 것 같습니다.
또한 HttpStatus CREATED를 활용하실거면 해당 CRATED을 찾아보고 맞춰서 필요한 정보 또한 제공해보면 어떨까요?

https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
https://developer.mozilla.org/ko/docs/Web/HTTP/Status/201

vsh123 replied:
현재 CommentApiController에 우선 적용했습니다. 확인 부탁드리겠습니다!

phs1116 commented on src/main/java/techcourse/fakebook/controller/ArticleApiController.java:
```diff
@@ -0,0 +1,51 @@
+import techcourse.fakebook.service.dto.UserOutline;
+
+@RestController
+@RequestMapping("/articles")
```
User와 관련된 Controller를 살펴보면 User에 대해서는  api인지 web인지에 따라 url path로 구분되어있는데 
Article과 Comment는 구분되어있지 않네요. :)
스펙을 UserController와 맞춰보면 어떨까요? 그 경우 어떤점에서 더 이득이 있을까요. :)

vsh123 replied:
현재 CommentApiController에 우선 적용했습니다. 확인 부탁드리겠습니다!

phs1116 commented on src/main/java/techcourse/fakebook/domain/article/Article.java:
```diff
@@ -0,0 +1,57 @@
+import javax.persistence.*;
+
+@Entity
+public class Article extends DateTime {
```
DateTime이라는 Abstract Class는 이를 상속한 객체들이 시간과 관련된 (is a 관계) 객체로 보입니다.
조금 더 적합한 추상 클래스 명을 선택하는게 좋을 것 같네요. :)

vsh123 replied:
BaseEntity로 수정하였습니다!

phs1116 commented on src/main/java/techcourse/fakebook/domain/article/Article.java:
```diff
@@ -0,0 +1,57 @@
+    private User user;
+
+    @Column(nullable = false)
+    private boolean isPresent;
```
boolean형에 대해서 변수명 자체는 형용사로, 메서드를 동사형으로 제공하는게 컨벤션입니다. :)
또한 일반적으로 present보다는 deleted와 같이 삭제라는 용어를 더 많이 사용하기도 해요.

phs1116 commented on src/main/java/techcourse/fakebook/domain/user/User.java:
```diff
@@ -0,0 +1,105 @@
+    private String name;
+
+    @Column(nullable = false)
+    private String gender;
```
gender가 도메인적으로 정의가 되어있지 않은 것 같네요.
성별이라는 것 자체가 남성/여성 과 같이 정해진 값이 존재할텐데요
enum을 통해 구체적으로 표현해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -0,0 +1,86 @@
+
+    Article getArticle(Long id) {
+        Article article = articleRepository.findById(id).orElseThrow(NotFoundArticleException::new);
+        if (article.isNotPresent()) {
```
모든 케이스에 대해서 존재하는것만 가져오기를 바란다면
JAP에  @Where 어노테이션이 있습니다. 이를 활용하거나
혹은 JPA 쿼리 자체에서 존재하는 값만 가져오도록 해도 좋을 것 같네요. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -0,0 +1,86 @@
+        article.delete();
+    }
+
+    Article getArticle(Long id) {
```
접근제어자에 대해서 package-private으로 정의한 이유가 있을까요?

vsh123 replied:
getArticle(), getUser()의 경우 다른 서비스에서 사용하는 경우가 있기 때문에 package-private으로 정의하였습니다!

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -0,0 +1,86 @@
+    public ArticleLikeResponse like(Long id, UserOutline userOutline) {
+        ArticleLike articleLike = new ArticleLike(userService.getUser(userOutline.getId()), getArticle(id));
+        if (articleLikeRepository.existsByUserIdAndArticleId(userOutline.getId(), id)) {
+            articleLikeRepository.delete(articleLike);
```
영속성 등록이 되지 않은 객체인데요, 정상적으로 삭제가 되나요?

vsh123 replied:
CommentLike에 대해서 삭제로직 우선 변경하였습니다!

phs1116 commented on src/main/java/techcourse/fakebook/service/CommentService.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private void checkAuthor(UserOutline userOutline, Comment comment) {
+        if (comment.getUser().isNotAuthor(userOutline.getId())) {
```
user가 author라는 Comment혹은 Article 도메인에 특화된 내용을 알고있는게 상당히 부자연스럽네요.
Getter를 활용하지말고 Comment와 메시지 전달을 통해서 확인하도록 구현해보면 어떨까요?

vsh123 replied:
수정완료했습니다!

phs1116 commented on src/main/java/techcourse/fakebook/domain/like/ArticleLike.java:
```diff
@@ -0,0 +1,28 @@
+import javax.persistence.*;
+
+@Entity
+public class ArticleLike {
```
대부분의 엔티티가 equals, hashcode, toString을 구현하지 않고 있네요. :)

vsh123 replied:
생성 완료했습니다!

phs1116 commented on src/main/java/techcourse/fakebook/domain/user/User.java:
```diff
@@ -0,0 +1,105 @@
+    private User() {
+    }
+
+    public User(String email, String encryptedPassword, String name, String gender, String coverUrl, String birth, String introduction) {
```
도메인에서 대부분 값들에 대한 유효성 체크가 없네요. 추가해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/dto/UserSignupRequest.java:
```diff
@@ -0,0 +1,62 @@
+package techcourse.fakebook.service.dto;
+
+public class UserSignupRequest {
+    private String email;
```
DTO에서도 모든 값 검증이 없는 것 같습니다.
프론트 혹은 API를 통해서 잘못된 값으로 요청이 들어와도 그대로 동작 할 것으로 보이는데요.
전체적으로 예외케이스에서 검증하고 validation을 추가하는게 좋을 것 같네요. :)

vsh123 replied:
validation 추가하였습니다!

phs1116 commented on src/main/java/techcourse/fakebook/service/UserService.java:
```diff
@@ -0,0 +1,77 @@
+        userRepository.delete(user);
+    }
+
+    @Transactional(readOnly = true)
```
readOnly에 대해서 분리가 필요하다면 CQRS를 적용하여 Query를 위한 서비스, Command를 위한 서비스를 분리할 수도 있습니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/dto/CommentLikeResponse.java:
```diff
@@ -0,0 +1,19 @@
+
+public class CommentLikeResponse {
+    private Long commentId;
+    private boolean isLiked;
```
존재하는 자원이 없을 경우에 isLiked = false로 전달하는데요
Rest API의 스펙에 맞게 HttpStatus로도 표현할 수 있을 것 같네요.

vsh123 replied:
수정 완료했습니다! 해당 부분이 의도해주신 피드백과 같은지는 잘모르겠습니다 ㅠㅠ 확인 후 리뷰 부탁드릴게요!

phs1116 replied:
네 케이스에 따른 Http 스펙으로의 전달 매우 잘하셨습니다. 👍

phs1116 commented on src/main/java/techcourse/fakebook/service/LoginService.java:
```diff
@@ -0,0 +1,47 @@
+        User user = userRepository.findByEmail(loginRequest.getEmail())
+                .orElseThrow(NotFoundUserException::new);
+
+        if (!encryptor.matches(loginRequest.getPassword(), user.getEncryptedPassword())) {
```
checkEncryptedPassword와 같은 로직인 것 같은데 두번 비교하는 이유가 있을까요?

eunsukko replied:
어느 곳에서 비밀번호를 검증해줄까 고민하다가 2가지 방식모두 적어놓고 까먹었습니다ㅠ!
감사합니다..!

phs1116 commented on src/test/java/techcourse/fakebook/controller/ArticleApiControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+    void 글을_잘_작성하는지_확인한다() {
+        ArticleRequest articleRequest = new ArticleRequest("hello");
+
+        given().
```
Rest Assured 깔끔하네요 :)

phs1116 commented on src/test/java/techcourse/fakebook/controller/ArticleApiControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+public class ArticleApiControllerTest extends ControllerTestHelper {
```
Article, Comment 둘 다 Like에 관련된 테스트가 없는 것 같습니다. 확인 부탁드릴게요. :)

vsh123 replied:
CommentApiController에 우선 추가완료했습니다!

phs1116 commented on src/test/java/techcourse/fakebook/controller/CommentApiControllerTest.java:
```diff
@@ -0,0 +1,79 @@
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+public class CommentApiControllerTest extends ControllerTestHelper {
```
CommentApiControllerTest, ArticleApiControllerTest 둘 다 CRUD 테스트 부분에 대해서 공통적인 부분이 있는데요.
기본적인 CRUD 테스트에서는 공통적으로 사용할 수 있도록 추상화해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/CommentService.java:
```diff
@@ -0,0 +1,98 @@
+package techcourse.fakebook.service;
```
domain은 각 도메인 별로 패키지가 분리되어있는데 
controller와 service는 그렇지 않네요.
통일성 있게 패키지를 분리해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/utils/UserAssembler.java:
```diff
@@ -0,0 +1,46 @@
+import techcourse.fakebook.service.utils.encryptor.Encryptor;
+
+@Component
+public class UserAssembler {
```
다른 Assembler는 static method로 제공하는데 UserAssembler만 빈으로 생성되고있네요.
통일성있게 맞추던가, 혹은 Encryptor 때문이라면 이에대한 의존을 분리하고 가져갈 수 있는 방법이 없는지 고민해보면 좋을 것 같습니다.

eunsukko replied:
static method 로 제공하면서 Encryptor 에 대한 의존성을 분리 할 방법이... 지금은 파라미터로 받는 방식만 생각이 납니다. (static 인 Encryptor 내부변수를 두고 나중에 추가해줄 것이 아니라면요)

이 경우에 결국은 UserAssembler를 사용하는 곳에서 사용할 Encryptor를 넣어주어야 하지요

결국 어디서 사용할 Encryptor 인스턴스를 찾아서 주입해주냐의 문제인 것 같은데...
아직까진 암호화관련 부분이 실행중에 다양해지지 않을 것 같아서
그런 면에서는 UserAssembler 내부에서 의존관계를 설정해주는게 낫지 않을까 싶습니다


파라미터로 의존을 받는 방식말고 static 에서 의존을 분리할 수 있는 다른 방식이 있을까요??

phs1116 replied:
말씀하신대로 빈이 아닌 객체가 빈 객체에 대한 의존을 가져가려면 파라미터로 받는 수밖에 없어요.
Encryptor부분이 다양해지지 않을 것이란 말씀은 Encryptor를 빈으로 설정할 필요가 없다는 말로도 받아들여져서 그냥 Encryptor도 static method로 가져가도 되지 않나 생각도 드네요. :)

혹은 나머지  Assembler도 모두 빈으로 등록하여 통일성 있게 맞추는것도 괜찮을 것 같아요.

phs1116 commented on src/test/java/techcourse/fakebook/service/ArticleServiceTest.java:
```diff
@@ -0,0 +1,76 @@
+    void 좋아요가_잘_취소되는지_확인한다() {
+        articleService.like(3L, userDto);
+        ArticleLikeResponse articleLikeResponse = articleService.like(3L, userDto);
+        assertThat(articleLikeResponse.isLiked()).isFalse();
```
response에 대해서 확인하지 실제로 값이 false로 내려오는지(변경되었는지)에 대한 테스트는 없는 것 같네요 :)

vsh123 replied:
해당 테스트 수정 완료했습니다!

phs1116 reviewed (changes requested):
안녕하세요. 리뷰어 화투입니다.
전체적으로 매우 잘 구현하셨습니다. 👍 
도메인, Validation 및 컨벤션, 테스트에 대해서 몇가지 수정되면 좋을 것 같은 점이 있어 피드백 추가했습니다.
참고 부탁드릴게요 :)

vsh123 committed:
docs : 1주차 피드백 Todo List 작성 (#19)

vsh123 committed:
[쉐이크반] refactor : Comment 리팩토링 (#27)

* feat : Comment CommentLike equals and hashcode

* refactor : CommentService.checkAuthor

* refactor : ApiController uri

* fix : CommentService like 수정

1. delete가 정상적으로 되지 않는 문제 해결

* refactor : ResponseEntity static method 사용

* refactor : comment like refactoring

1. HttpStatus return으로 변경
2. Controller 테스트 코드 추가

* refactor : User.isNotAuthor() renaming

vsh123 committed:
[쉐이크반] refactor : Common refactoring (#23)

* rename : DateTime Entity 이름 변경

* refactor : ArticleAssembler, CommentAssembler Bean 설정

vsh123 committed:
fix : bug fix

1. js uri수정
2. 로그인 오류 수정

vsh123 commented:
안녕하세요! 
보내주신 피드백 및 프론트 엔드 작업하여 리뷰 요청 드립니다!
이번 작업에서 다음과 같은 사항에 대해 궁금한점이 생기는데 의견 주시면 감사하겠습니다!

## back-end
1. 말씀해주신 레이어 안에서 도메인별로 패키지를 분리하려고 하니까 패키지가 너무 많아지는 느낌이 들었습니다.
또한 domain 안에서는 각 도메인 패키지별로 명확하게 파일들을 분리할 수 있었지만 service와 controller의 경우에는 utils나 dto같은 파일들이 있는데 이런 것들을 어떻게 분리해야 할지도 감이 잡히지 않습니다 ㅠㅠ 혹시 이에 대해 참고할만한 가이드가 있으면 부탁드리겠습니다..!

피드백 반영한 부분은 따로 댓글을 남겼으며, 아직 반영하지 못한 부분은 최대한 빨리 수정해서 push하도록 하겠습니다!

vsh123 committed:
fix : front BASE_URL 수정

woowahan-pjs commented on src/main/java/techcourse/fakebook/service/LoginService.java:
```diff
@@ -0,0 +1,47 @@
+            throw new NotMatchPasswordException();
+        }
+
+//        String encryptedPassword = encryptor.encrypt(loginRequest.getPassword());
```
주석 제거!

woowahan-pjs replied:
주석 제거!

woowahan-pjs reviewed (approved):
당장 `controller`와 `service`의 패키지를 나누기 어렵다면
클래스 이름을 잘 살펴보고 그것을 기준으로 나눠 볼 수 있어요.
하지만 `controller.utils`는 어디에 둘 건지, `utils`가 적절한지 고민해 보세요.
아래 글은 가볍게 참고해 주세요.
https://www.slipp.net/questions/36
